src/main/scala/lms/core/types.scala

```diff
@@ -0,0 +1,147 @@
+package lms.core
+
+import lms.core.Backend.{Const, Exp, Node, Sym}
+import lms.core.TypeMap.TypeExp
+
+import scala.collection.mutable
+
+
+object Reflect {
+  def unapply(x: Any)(implicit tm: TypeMap): Option[(String, List[Exp])] = x match {
+    case s@Sym(_) =>
+      tm.g.findDefinition(s).map(n => (n.op, n.rhs.filter(_.isInstanceOf[Exp]).map(_.asInstanceOf[Exp])))
+    case _ => None
+  }
+}
+
+abstract class Type[T] {
+  def reify(x: Exp, tm: TypeMap): TypeExp
+}
+
+abstract class SimpleType[T] extends Type[T] {
+  def reify(x: Exp, tm: TypeMap): TypeExp = reify(tm)
+  def reify(tm: TypeMap): TypeExp
+}
+
+object TypeExp {
+  def of[T](x: Exp)(implicit t: Type[T], typeMap: TypeMap): TypeExp = t.reify(x, typeMap)
+}
+
+object TypeNames extends Enumeration {
+  val Unit, Boolean, Char, Short, Int, Float, Double = Value
+}
+
+// Primitive types corresponding to Scala ones
+object Types {
+
+  val Unit: TypeExp = Const(TypeNames.Unit)
+  val Boolean: TypeExp = Const(TypeNames.Boolean)
+  val Char: TypeExp = Const(TypeNames.Char)
+  val Short: TypeExp = Const(TypeNames.Short)
+  val Int: TypeExp = Const(TypeNames.Int)
+  val Float: TypeExp = Const(TypeNames.Float)
+  val Double: TypeExp = Const(TypeNames.Double)
+
+
+  object Array {
+    val name = "Array"
+
+    def apply(et: TypeExp)(implicit tm: TypeMap): TypeExp = tm.g.reflect(Array.name, et)
+
+    def unapply(x: Any)(implicit tm: TypeMap): Option[TypeExp] = x match {
+      case Reflect(Array.name, t :: Nil) => Some(t)
+    }
+  }
+
+}
+
+case class TypeMap(g: GraphBuilder) extends mutable.HashMap[Exp, TypeExp]
+
+object TypeMap {
+
+  type TypeExp = Exp
+  def apply(): TypeMap = {
+    // TODO:  should we have the same GraphBuilder for types
+    //        and backend or can we refactor it
+    //        to share some code?
+    val g = new GraphBuilder
+    // TODO:  why is `g` initialized without any `curLocalDefs`?
+    //        What about `curLocalReads` etc which are not used yet.
+    g.curLocalDefs = Set()
+    TypeMap(g)
+  }
+
+}
+
```

Comment:
The current `typeMap` is part of the global state in `Adapter`. The new version of `TypeMap` probably won't have a `GraphBuilder` and will use the one that is used for reflecting terms `Adapter.g`.

Reply:
A GraphBuilder object is not long-living. After a Graph is constructed, the GraphBuilder object is used to generate a Graph object, and then discarded. If we have a HashMap in the GraphBuilder, we might also need to save it to the Graph object. It might not be a bad idea.

So far we have been using Metadata to track additional information for Graph. When doing transformation, the handling of Metadata gets hacky (using Adapter.oldTypeMap, for instance). If this Metadata is part of the Graph, then that looks cleaner imo.